airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/base/IntegrationRunner.java

```diff
@@ -197,31 +206,55 @@ private void produceMessages(final AutoCloseableIterator<AirbyteMessage> message
   }
 
   @VisibleForTesting
-  static void consumeWriteStream(final AirbyteMessageConsumer consumer) throws Exception {
-    // use a Scanner that only processes new line characters to strictly abide with the
-    // https://jsonlines.org/ standard
-    final Scanner input = new Scanner(System.in, StandardCharsets.UTF_8).useDelimiter("[\r\n]+");
-    consumer.start();
-    while (input.hasNext()) {
-      consumeMessage(consumer, input.next());
+  static void consumeWriteStream(final SerializedAirbyteMessageConsumer consumer) throws Exception {
+    try (final BufferedInputStream bis = new BufferedInputStream(System.in);
+        final ByteArrayOutputStream baos = new ByteArrayOutputStream()) {
+      consumeWriteStream(consumer, bis, baos);
     }
   }
 
-  private static void runConsumer(final AirbyteMessageConsumer consumer) throws Exception {
-    watchForOrphanThreads(
-        () -> consumeWriteStream(consumer),
-        () -> System.exit(FORCED_EXIT_CODE),
-        INTERRUPT_THREAD_DELAY_MINUTES,
-        TimeUnit.MINUTES,
-        EXIT_THREAD_DELAY_MINUTES,
-        TimeUnit.MINUTES);
+  @VisibleForTesting
+  static void consumeWriteStream(final SerializedAirbyteMessageConsumer consumer,
+                                 final BufferedInputStream bis,
+                                 final ByteArrayOutputStream baos)
+      throws Exception {
+    consumer.start();
+
+    final byte[] buffer = new byte[8192]; // 8K buffer
+    int bytesRead;
+    int byteCount = 0;
+    boolean lastWasNewLine = false;
+
+    while ((bytesRead = bis.read(buffer)) != -1) {
+      for (int i = 0; i < bytesRead; i++) {
+        final byte b = buffer[i];
+        if (b == '\n' || b == '\r') {
+          if (!lastWasNewLine && baos.size() > 0) {
```

Comment:
what's the difference between byteCount and baos.size()?

Reply:
* `bytesRead` tracks the bytes read out of the buffer. i.e. (1 of 8192)
* `byteCount` is tracking number of bytes read for a single message.

Reply:
sorry, I was more thinking about bytesCount vs the `baos.size()` call. I think we don't need bytesCount at all, and can rely on the bytearrayoutputstream to track it for us?

(though obv in the other comment chain, I'm trying to find a way to avoid this entirely)

Reply:
agreed.